models/dpr_test/dpr_test.c

```diff
@@ -0,0 +1,239 @@
+  fe_dpr_test_devp = devm_kzalloc(&pdev->dev, sizeof(fe_dpr_test_dev_t), GFP_KERNEL);
+  r = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+  if (r == NULL) {
+    pr_err("IORESOURCE_MEM (register space) does not exist\n");
+    goto bad_exit_return;  }
+  fe_dpr_test_devp->regs = devm_ioremap_resource(&pdev->dev, r);
+  if (IS_ERR(fe_dpr_test_devp->regs)) {
+    ret_val = PTR_ERR(fe_dpr_test_devp->regs);
+    goto bad_exit_return;
+  }
+  platform_set_drvdata(pdev, (void *)fe_dpr_test_devp);
+  fe_dpr_test_devp->name = devm_kzalloc(&pdev->dev, 50, GFP_KERNEL);
+  if (fe_dpr_test_devp->name == NULL)
+    goto bad_mem_alloc;
+  strcpy(fe_dpr_test_devp->name, (char *)pdev->name);
+  pr_info("%s\n", (char *)pdev->name);
+  status = alloc_chrdev_region(&dev_num, 0, 1, "fe_dpr_test_");
+  if (status != 0)
+    goto bad_alloc_chrdev_region;
+  sprintf(deviceMinor, "%d", MAJOR(dev_num));
+  strcat(device_name, deviceMinor);
+  pr_info("%s\n", device_name);
+  cl = class_create(THIS_MODULE, device_name);
+  if (cl == NULL)
+    goto bad_class_create;
+  cdev_init(&fe_dpr_test_devp->cdev, &fe_dpr_test_fops);
+  status = cdev_add(&fe_dpr_test_devp->cdev, dev_num, 1);
+  if (status != 0)
+    goto bad_cdev_add;
+ device_obj = device_create_with_groups(cl, NULL, dev_num, NULL, dpr_test_groups, device_name);
+  if (device_obj == NULL)
+    goto bad_device_create;
+  dev_set_drvdata(device_obj, fe_dpr_test_devp);
+  pr_info("dpr_test exit\n");
+  return 0;
+bad_device_create:
+bad_cdev_add:
+  cdev_del(&fe_dpr_test_devp->cdev);
+bad_class_create:
+  class_destroy(cl);
+bad_alloc_chrdev_region:
+  unregister_chrdev_region(dev_num, 1);
+bad_mem_alloc:
+bad_exit_return:
+  pr_info("dpr_test_probe bad exit\n");
+  return ret_val;
+}
+
+
+static int dpr_test_open(struct inode *inode, struct file *file) {
+  // TODO: fill this in (if its needed, it might not be)
+  return 0;
+}
+
+static int dpr_test_release(struct inode *inode, struct file *file) {
+  // TODO: fill this in (if its needed, it might not be)
+  return 0;
+}
+
+static ssize_t dpr_test_read(struct file *file, char *buffer, size_t len, loff_t *offset) {
+  // TODO: fill this in (if its needed, it might not be)
+  return 0;
+}
+
+static ssize_t dpr_test_write(struct file *file, const char *buffer, size_t len, loff_t *offset) {
+  // TODO: fill this in (if its needed, it might not be)
+  return 0;
+}
+static ssize_t name_read(struct device *dev, struct device_attribute *attr, char *buf) {
+  fe_dpr_test_dev_t * devp = (fe_dpr_test_dev_t *)dev_get_drvdata(dev);
+  sprintf(buf, "%s\n", devp->name);
+  return strlen(buf);
+}
+
+static ssize_t dpr_data_read(struct device *dev, struct device_attribute *attr, char *buf) {
+  fe_dpr_test_dev_t * devp = (fe_dpr_test_dev_t *)dev_get_drvdata(dev);
+  fp_to_string(buf, devp->dpr_data, 0, false, 32);
+  strcat2(buf,"\n");
+  return strlen(buf);
+}
+
+static ssize_t dpr_data_write(struct device *dev, struct device_attribute *attr, const char *buf, size_t count) {
```

Comment:
What is the interface for writing to dpr_data in sysfs? It seems like it is a string of numbers separated by spaces or commas? Where does `count` come in?

Reply:
Yep, that's the basic format.  Count comes from the signature the kernel expects for the platform device.

Reply:
>  Count comes from the signature the kernel expects for the platform device

Can you elaborate on that? I have no idea what that means in practice. Where does the signature come from?

Reply:
Also, `count` is the number of words/addresses, right?

Reply:
Essentially.  More info can be found [here](https://www.kernel.org/doc/Documentation/filesystems/sysfs.txt)

Reply:
I'm even more confused now. `count` is the reference count for the kobject? Can you please just summarize it in laymen's terms? I.e. how does writing a string of numbers to `dpr_data` actually map to a DPRAM of a given size?

Reply:
Count is really just the number of characters passed in the buffer. And when writing to the driver, they can be comma separated, space separated, or even on line breaks due to the implementation. The driver does expect that you will write the number of values needed to update the dual port ram in one big write.

Reply:
@ fe-wickham That makes sense. I was reading the for loop wrong. I had it in my head that count was being used after splitting the string into substrings, but I see that is not the case. I just can't read.